xcp_d/interfaces/regression.py

```diff
@@ -53,43 +54,88 @@ def _run_interface(self, runtime):
 
         # Get the confound matrix
         confound = pd.read_table(self.inputs.confounds)
+        confound_arr = confound.to_numpy()
+
+        # Any columns starting with "signal__" are assumed to be signal regressors
+        signal_columns = [c for c in confound.columns if c.startswith("signal__")]
+        if signal_columns:
+            LOGGER.info(
+                "Performing nonaggressive denoising using the following signal columns: "
+                f"{', '.join(signal_columns)}"
+            )
+            noise_columns_idx = [
+                i for i, c in enumerate(confound.columns) if c not in signal_columns
+            ]
 
-        confound = confound.to_numpy().T  # Transpose confounds matrix to line up with bold matrix
         # Get the nifti/cifti matrix
-        bold_matrix = read_ndata(datafile=self.inputs.in_file, maskfile=self.inputs.mask)
+        bold_arr = read_ndata(datafile=self.inputs.in_file, maskfile=self.inputs.mask)
+        bold_arr = bold_arr.T  # transpose BOLD data to TxS
+
+        # Regress out the confounds via linear regression from sklearn
+        if bold_arr.shape[0] < confound_arr.shape[1]:
+            LOGGER.warning(
+                "Warning: Regression might not be effective due to rank deficiency, "
+                "i.e., the number of volumes in the bold file is smaller than the number "
+                "of regressors."
+            )
 
-        # Demean and detrend the data
+        if signal_columns:
+            # Perform non-aggressive denoising.
+            # First, mean-center and detrend BOLD data
+            bold_arr = signal.clean(
+                signals=bold_arr,
+                detrend=True,  # this mean-centers and linearly detrends the data
+                standardize=False,
+                confounds=None,
+                filter=None,
+                ensure_finite=True,
+            )
```

Comment:
Should we also detrend the confounds?

[Per Nilearn](https://github.com/nilearn/nilearn/blob/cfcebceba4f180a7f55f9553ab1d73478961d5b8/nilearn/signal.py#L629-L631) (citing [Lindquist et al. [2019]](https://doi.org/10.1002/hbm.24528)):

>Detrend and filtering should apply to confounds, if confound[s] present
>keep filters orthogonal (according to Lindquist et al. (2018))
>Restrict the signal to the orthogonal of the confounds

Reply:
Per Matt, since we filter _after_ regression, it should be fine. While detrending and then regressing out undetrended confounds can reintroduce trends, those trends then get removed by the filtering done later.

On the other hand... users can disable bandpass filtering, so we should probably do the regression properly.

Reply:
Confounds will now be mean-center and detrended in the same manner as the BOLD data.